[BB-3174] Automate estimation session

Agrendalath opened the pull request:
This adds estimation session automation for the [Agile Poker](https://marketplace.atlassian.com/apps/700473/agile-poker-for-jira-planning-estimation) Jira app.

# Prerequisites
1. Set `.env` (you can find them [here](https://vault.opencraft.com:8200/ui/vault/secrets/secret/show/core/Sprints%20Development%20Envs)). There were some recent changes.
1. To `sprints/dashboard/tasks.py:update_estimation_session_task`, add the following below `for cell in get_cells(conn):`:
	```python
	if cell.board_id != 26:
	    continue
	```
1. To `sprints/dashboard/tasks.py:update_estimation_session_task`, add the following below `all_member_keys = list(current_member_keys | member_keys)`:
	```python
	print(f"Members of {cell.name}: {all_member_keys}")
	print(f"Unestimated tickets from {cell.name}: {[conn.issue(issue.id).key for issue in issues if issue.key.startswith(cell.key)]}")
	```
1. To `sprints/dashboard/tasks.py:move_estimates_to_tickets_task`, add the following below `for cell in get_cells(conn):`:
	```python
	if cell.board_id != 26:
	    continue
	```
1. To `sprints/dashboard/tasks.py:move_estimates_to_tickets_task`, add the following below `final_vote = get_poker_session_final_vote(votes, vote_values)`:
	```python
	print(f"The final vote for {conn.issue(issue).key} is {final_vote}.")
	```
1. To `sprints/dashboard/automation.py:ping_users_on_ticket`, add the following above `if not settings.DEBUG:`:
	```python
	print(message)
	```

# Testing instructions
1. Start the backend with `docker-compose -f local.yml up --build`.
1. Run migrations with `docker-compose -f local.yml exec django python manage.py migrate`.
1. Open Django shell with `docker-compose -f local.yml exec django python manage.py shell_plus` and run:
	```python
	from sprints.dashboard.tasks import *
	from sprints.dashboard.libs.jira import *
	
    conn = CustomJira(
        server=settings.JIRA_SERVER,
        basic_auth=(settings.JIRA_USERNAME, settings.JIRA_PASSWORD),
        options={
            'agile_rest_path': GreenHopperResource.AGILE_BASE_REST_PATH,
            'headers': {
                'Referer': 'https://tasks.opencraft.com/download/resources/com.spartez.jira.plugins.jiraplanningpoker/frontend/index.html',
            }
        }
    )
	```

1. To test creating estimation session, use the following snippet:
	```python
	session = conn.create_poker_session(
	    board_id=26,  # This is the Bebop one.
	    name=get_next_poker_session_name(conn),
	    issues=[],
	    participants=['viadanna'],
	    scrum_masters=[conn.myself()['key']],
	    send_invitations=False,
	)
	```
1. Visit the "Estimates" section in Jira and see that the session called "Sprint 241 & Stretch Goals" has been created. It should have only one button - "Estimate". Open the session and see that it does not work - the "Loading issues..." is there all the time. This is intended behavior, as explained in the `create_estimation_session_task`'s docstring.
1. Run the following snippet:
	```python
	session.update({'issuesIds': ['27146'], 'scrumMasters': ['crafty', 'viadanna']})
	```
1. Reload the page and check that `BB-1986` has been added there. Also, check that you can click on the `··· -> Configure session` button, as you are the scrum master of this session.
1. To test updating estimation session task, run the following snippet and check that the output looks reasonable:
	```python
	update_estimation_session_task()
	```
1. To verify retrieving session for a cell, run the following snippet and see that it does not raise any exception:
	```python
	assert len(conn.poker_sessions(26, state="OPEN", name=get_next_poker_session_name(conn))) == 1
	assert session.sessionId == conn.poker_sessions(26, state="OPEN", name=get_next_poker_session_name(conn))[0].sessionId
	```
1. Close the session with the following and verify that it's closed in Jira.
	```python
	conn.close_poker_session(session.sessionId, False)
	```
1. Run the following and check that it displays a message with a comment to me that it couldn't estimate the issue.
	```python
	move_estimates_to_tickets_task()
	```
1. Use `··· -> Restart session` in Jira to reopen the session (do not add any new tickets, just click `Next`). Vote for the ticket on the session page.
1. Run the following and check that it displays a message with the estimate you have submitted.
	```python
	move_estimates_to_tickets_task()
	```
1. Clean up by running the following:
	```python
	session.delete()
	```

Agrendalath commented on sprints/dashboard/utils.py:
```diff
@@ -240,16 +240,16 @@ def get_sprints(conn: CustomJira, board_id: int) -> List[Sprint]:
 
 
-def get_sprint_number(previous_sprint: Sprint) -> int:
+def get_sprint_number(sprint: Sprint) -> int:
```
This is not related, but the wording of this function was confusing - its purpose has changed after a refactor, but we forgot to rename internal variables.

viadanna commented on sprints/dashboard/tasks.py:
```diff
@@ -413,3 +434,146 @@ def unflag_tickets_task() -> None:
+                members = get_cell_member_names(conn, get_cell_members(conn.quickfilters(cell.board_id)))
+                member_keys = set(user.key for user in all_users if user.displayName in members)
+                current_member_keys = set(user.userKey for user in poker_session.participants)
+                all_member_keys = list(current_member_keys | member_keys)
```
This means it doesn't make sense to remove newcomers after the Poker session has them, as they will be set again, right? Maybe add this to the docstring?

Agrendalath replied:
Good idea, I added it to the docstring and the docs.

viadanna commented on sprints/dashboard/tasks.py:
```diff
@@ -413,3 +434,146 @@ def unflag_tickets_task() -> None:
+        - estimation,
+        - configuration.
+    Therefore, the decision is to avoid adding the participants to the session until there are issues that can be added
+    too. Assuming that the sessions are fully automated, and don't require any manual interventions in the beginning,
```
👍

viadanna commented on sprints/dashboard/tasks.py:
```diff
@@ -413,3 +434,146 @@ def unflag_tickets_task() -> None:
+
+        for cell in get_cells(conn):
+            try:
+                poker_session = conn.poker_sessions(cell.board_id, state="CLOSED", name=session_name)[0]
```
Is this guaranteed to return the newest one? Sorting can ensure that :)

Agrendalath replied:
Good point, though the API response for sessions does not provide any meaningful data we could sort them by. Therefore I've just changed this approach to use all sessions with this name, just like the `close_estimation_session_task` does.

viadanna commented on sprints/dashboard/tasks.py:
```diff
@@ -413,3 +434,146 @@ def unflag_tickets_task() -> None:
+            try:
+                poker_session = conn.poker_sessions(cell.board_id, state="CLOSED", name=session_name)[0]
+            except IndexError:
+                if not settings.DEBUG:
```
This will probably be triggered when a new cell is created and no estimation session exists yet, right? A FIXME might be a good idea here ;)

Agrendalath replied:
Yep, it's good to mention the use case for this. I have added it as a regular comment instead of a `FIXME`, but I have expanded the logged error to indicate why this might have happened and how to fix it.

viadanna reviewed (commented):
@Agrendalath Looking good, just a few comments.

I couldn't create a session, it returns 403. I've updated my env, do you know what I'm missing?

Agrendalath commented:
@viadanna,
> I couldn't create a session, it returns 403. I've updated my env, do you know what I'm missing?

Ooops, I forgot to update the testing instructions with this ugly `Referer` header for the Jira connection (I'm just using pre-made snippets for testing these things locally, so it worked fine for me). I updated this now.

Agrendalath commented on sprints/dashboard/tasks.py:
```diff
@@ -413,3 +434,174 @@ def unflag_tickets_task() -> None:
+        session_name = get_next_poker_session_name(conn)
+        issues = get_unestimated_next_sprint_issues(conn)
+        # FIXME: This will not work for more than 1000 users. To support it, add `startAt` to handle the pagination.
+        all_users = conn.search_users("''", maxResults=1000)  # Searching for the "quotes" returns all users.
```
Yay, weird API design, which is `Gathering Interest` [since 2012](https://jira.atlassian.com/browse/JRASERVER-29069?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&showAll=true).

viadanna commented on README.rst:
```diff
@@ -430,6 +430,10 @@ Updating sessions
 ~~~~~~~~~~~~~~~~~
 This task adds any tickets that have been added to the next sprint but are not present in the estimation session. It also adds participants, when there are tickets scheduled for the next sprint (please see the explanation above), or if a new member joins a cell.
 
+Note
```
👍

viadanna commented on sprints/dashboard/tasks.py:
```diff
@@ -476,8 +476,10 @@ def update_estimation_session_task() -> None:
-    This does not override the manual changes made to the session - i.e. if issue or user has been added manually to the
+    This does not override the manual additions to the session - i.e. if an issue or user has been added manually to the
     session, then it will be retained, as it merges available issues and participants with the applied ones.
+    However, any removed items (e.g. an issue scheduled for the next sprint, or a user who is a member of the cell)
```
👍

viadanna commented on sprints/dashboard/tasks.py:
```diff
@@ -486,12 +488,30 @@ def update_estimation_session_task() -> None:
+                poker_session = conn.poker_sessions(cell.board_id, state="OPEN", name=session_name)[0]
+            except IndexError:
+                if not settings.DEBUG:
+                    # It can happen:
```
👍

viadanna commented on sprints/dashboard/tasks.py:
```diff
@@ -548,32 +568,40 @@ def move_estimates_to_tickets_task() -> None:
+            poker_sessions = conn.poker_sessions(cell.board_id, state="CLOSED", name=session_name)
+
+            if not poker_sessions and not settings.DEBUG:
+                # This can happen if a new cell has been added, then its session was created manually, and it either:
```
👍

viadanna reviewed (approved):
Excellent work @Agrendalath

Good to go 👍

- [x] I tested this locally.
- [x] I read through the code.

0x29a commented on config/settings/base.py:
```diff
@@ -556,19 +560,45 @@
+    # FIXME: The sprint completion is not automated yet, and this should be done only if the sprints all cells are
+    #  completed for all cells. Add this to the sprint completion pipeline, once it's automated.
+    "sprints.dashboard.tasks.create_estimation_session_task": {
+        "name": "[ASYNC] Close estimation session",
```
@Agrendalath, shouldn't this be `Create estimation session`?

Agrendalath replied:
@0x29a, yep, good catch. I addressed this in #70.
